### PR TITLE
docs(rendering): document incremental renderer contract

### PR DESCRIPTION
**Pull request by paulfalgout**
## Linked issue

- No linked issue. This completes the pre-beta renderer-contract documentation item in `ROADMAP.md`.

## What

- Document that custom renderers run with the View as `this` and may return `undefined` after committing directly to `view.el`.
- Update the Morphdom example to preserve the View root and skip identical subtrees with `isEqualNode`.
- Add an integration regression for renderer context and the direct-commit path.

## Why

The existing renderer already supports incremental DOM libraries, but the guide only implied the relevant contract. Making first-render versus update behavior explicit lets applications evaluate Morphdom, Lit, and similar optional renderers without adding a new core API before v5 beta.

## Agent-development failure addressed

- The prior guide described the renderer arguments but did not state its `this` binding or the exact `undefined` sentinel.
- Its Morphdom example replaced the View root on first render and traversed identical subtrees on updates.

## Public behavior contract

- Canonical behavior after this PR: a renderer receives `(template, data)` as a View method; any result other than `undefined` is passed to `attachElContent`, while `undefined` means the renderer already committed the update.
- Superseded behavior removed, if any: the stale root-replacing Morphdom example and pluralized `attachElContents` documentation are removed.
- Compatibility exception and removal condition, if any: none. Production behavior is unchanged.

## Runtime cost

- [x] Static or documentation only
- [ ] Development or test only; excluded from production entrypoints
- [ ] Changes an existing production path
- [ ] Adds an explicitly opt-in runtime path

Measured impact or explanation:

- Bundle: no production source or distribution changes.
- Startup/hot path: unchanged.
- Allocations/retention: unchanged.

## Scope

- Affects the View rendering guide, its roadmap status, and one View renderer integration test.
- Does not add a renderer argument, renderer package, Lit dependency, automatic attribute refresh, or production runtime behavior.
- Benchmark implementations and measurements remain outside this PR.

## Deployment Impact

- No runtime package or application redeploy is required.
- The documentation site will publish through the normal post-merge documentation workflow.

## Validation

- [x] Tests cover the acceptance criteria and relevant edge cases.
- [x] Docs, examples, declarations, diagnostics, and rule metadata agree where relevant.
- [x] No private consumer, private fixture, or undocumented context is required.
- [x] I ran the commands listed below and am reporting their actual results.

Commands:

```sh
npm run lint:ci
npm run coverage
npm run docs:check
npm run test:source
git diff --check
```

## Package and browser impact

- Entry points or install fixtures affected: none.
- Browser contracts affected: none; the existing renderer contract is newly explicit and regression-covered in JSDOM.
- Production/dev/test module-graph impact: production and development graphs are unchanged; one existing unit suite gains a test.

## Rollback or deprecation

- Revert this documentation/test commit. There is no runtime state or compatibility migration.

## Testing

- Full lint passed.
- 111 test files and 2,056 tests passed with 100% statements, branches, functions, and lines.
- All 19 agent benchmark contract tests passed.
- 23 executable documentation examples, 56 generated HTML files, and 503 internal links passed validation.
- Source validation and `git diff --check` passed.

## Review notes

- The example deliberately returns HTML on first render and directly patches on later renders.
- The `isEqualNode` bailout is documented for measurement; this PR does not claim benchmark improvement.
- Lit is identified only as a consumer of the existing contract, not as a core dependency or official package commitment.

---
## Summary by cubic
Documents the renderer contract for incremental DOM libraries like `morphdom` and `Lit`. Production behavior is unchanged; previously the guide only implied the contract and its `morphdom` example replaced the View root on first render.

**Docs**
- Renderers run with the View as `this`; returning `undefined` skips `attachElContent` and signals a direct commit.
- Updated the `morphdom` example to preserve the View root and skip identical subtrees with `isEqualNode`.
- Marked the renderer-contract roadmap item as done.

**Testing**
- Added a regression test for renderer context and the direct-commit path.

<sup>Written for commit a4c3ec9499f4e277192ce8ea4cca63d4f82fdfde. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/marionettejs/marionette/pull/396?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by paulfalgout** [comment]
**Approved** — reviewed at fetched SHA `9345cb26` (base `e6dd9e72`). Docs and one spec; no distribution change, so the gated aggregate stays at **90,127 / 100,000**.

I verified both documented claims against the implementation rather than taking them:

- `setRenderer` assigns `this.prototype._renderHtml = renderer` (`runtime/renderer.js`), and `_renderTemplate` calls `this._renderHtml(template, data)` — so the View **is** the `this` binding for a configured renderer, not only for the default method.
- `if (typeof html !== 'undefined') { this.attachElContent(html); }` — `undefined` alone suppresses the attach. Worth noting the doc's phrasing, *"anything other than `undefined`"*, is exactly right: `null` is still passed through, which a looser "returns nothing" would have obscured.

### The `attachElContents` correction is the most valuable fix here

The pluralized name appeared in the prose, the section heading, and the "you can modify… by overriding" sentence. The real method is `attachElContent`, singular.

That is worse than an ordinary typo because it named an **override point**. A reader following the old documentation would have defined `attachElContents` on their View, seen no error, and watched the default `attachElContent` keep running — output looks correct, the override silently does nothing, and the only symptom is that a customization had no effect. Fixing the heading matters as much as the prose, since that is what someone lands on from a search.

### The Morphdom example was actively wrong, and the old text admitted it

> *(Note in this case the view's `el` created at instantiation would be overridden)*

Replacing `view.el` breaks delegated events, Region ownership, and attachment tracking — the parenthetical acknowledged the behaviour without flagging it as a problem. The rewritten example patches the existing root instead, which is what a renderer contract designed around `this.el` should demonstrate.

Adding *"use a regular function when the renderer needs access to the View instance"* closes a real modern-JS trap: `setRenderer` installs onto the prototype and relies on call-site binding, so an arrow function passed there captures its lexical `this` and silently loses the View.

### Roadmap status

Moving from *"Before beta, document that contract explicitly"* to *"The first-render/update contract and Morphdom's `isEqualNode` subtree bailout are documented explicitly"* while keeping the measurement work outstanding is correct partial completion — the item narrows rather than closing, so the Lit and bailout experiments stay on the books.

File: docs/view.rendering.md [hunk 1]
```diff
@@ -220,36 +228,45 @@ attachElContent(html) {
...
 
-In this case because the renderer is modifying the `el` directly, there is no need to return the result
-of the template rendering for the view to handle in [`attachElContents`](#customizing-attachelcontents).
-It is certainly an option to return the compiled DOM and modify [`attachElContents`](#customizing-attachelcontents)
-to handle a DOM object instead of a string literal, but in many cases it may be overcomplicated to do so.
+The root element remains owned by the View. Other incremental renderers, such
+as Lit, can use the same contract: commit within `this.el` and return
+`undefined`. Root-level declarations remain separate; reevaluate a dynamic
+`className`, `id`, or `attributes` definition with
+[`renderAttributes()`](./marionette.view.md#renderattributes).
```
[the remark this answers is not in the document]
Comment (paulfalgout) [reply docs/view.rendering.md]: Fixed at amended head `a4c3ec9499f4e277192ce8ea4cca63d4f82fdfde`; the link now targets `#refreshing-root-attributes`, and `npm run docs:check` passes.

**Comment by paulfalgout** [comment]
**Approval stands** at `a4c3ec94`. Docs only; aggregate remains **90,127**.

Verified the anchor rather than assuming: `docs/marionette.view.md:86` is `### Refreshing Root Attributes`, which slugs to `#refreshing-root-attributes`, and no `#renderattributes` link survives anywhere under `docs/`.

Worth noting this was a link to a section added in #395 one PR ago — the kind of cross-document reference that goes stale precisely when two PRs land close together, and the sort of thing `docs:check`'s internal-link validation exists to catch. Good that it did.

**Comment by paulfalgout** [comment]
**Approval stands** at exact head `a4c3ec9499f4e277192ce8ea4cca63d4f82fdfde` against base `e6dd9e720db58cb5f8ea34d5cec12e0da68ca688`.

Cubic’s valid fragment-link finding is fixed, replied to, and resolved. Exact-head CI, documentation, Node 26 advisory, CodeQL, bundle-size, macOS/Windows package smoke, and reporting-only performance checks pass. The production artifacts remain unchanged at `90,127 / 100,000` Brotli-11 bytes, and 0 of 1 review threads are unresolved. CodeRabbit and Copilot were rate/quota limited and opened no threads.
